extension-workspace-references.md

```diff
@@ -0,0 +1,113 @@
+
+  /**
+   * The server provides workspace references exporting support.
+   */
+  xworkspaceReferencesProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace references request is sent from the client to the server to locate project-wide references to a symbol given its description / metadata.
+
+_Request_
+* method: 'workspace/xreferences'
+* params: `WorkspaceReferencesParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace References Request.
+ */
+interface WorkspaceReferencesParams {
+    /**
+     * Metadata about the symbol that is being searched for.
+     */
+    query: SymbolDescriptor;
+
+    /**
+     * An optional list of files to restrict the search to.
+     */
+    files?: string[];
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces, etc.
+ */
+interface ReferenceInformation {
+    /**
+     * The location in the workspace where the `symbol` is referenced.
+     */
+    reference: Location;
+
+    /**
+     * Metadata about the symbol that can be used to identify or locate its
+     * definition.
+     */
+    symbol: SymbolDescriptor;
+}
+```
+* error: code and message set in case an exception happens during the workspace references request.
+
+Where `SymbolDescriptor` is defined as follows:
+
+```typescript
+/**
+ * Represents information about a programming construct that can be used to
+ * identify and locate the construct's symbol. The identification does not have
+ * to be unique, but it should be as unique as possible.
+ *
+ * In contrast to `SymbolInformation`, `SymbolDescriptor` includes more concrete,
+ * language-specific, metadata about the symbol.
+ */
+interface SymbolDescriptor {
+    /**
+     * A list of properties of a symbol that can be used to identify or locate
+     * it.
+     */
+    [attr: string]: any
+}
+```
+
+### Goto Definition Extension Request
+
+This method is the same as `textDocument/definition`, except that:
+
+1. The method returns metadata about the definition (the same metadata that `workspace/xreferences` searches for).
+2. The concrete location to the definition (`location` field) is optional. This is useful because the language server might not be able to resolve a goto definition request to a concrete location (e.g. due to lack of dependencies) but still may know _some_ information about it.
+
+_Request_
+* method: 'textDocument/xdefinition'
+* params: [`TextDocumentPositionParams`](#textdocumentpositionparams)
+
+_Response_:
+* result: `LocationInformation[]` defined as follows:
+```typescript
+interface LocationInformation {
+    /* The location where the symbol is defined, if any. */
+    location?: Location;
+
+    /**
+     * Metadata about the symbol that can be used to identify or locate its
+     * definition.
+     */
+    symbol: []SymbolDescriptor;
```

Comment:
Well a package not published on NPM can only be installed from GitHub (or another VCS). In that case package.json would contain
```json
"facebook/react": "bd425837a4b5e45208e8944257ad40c9d08f04e5"
```
instead of
```json
"react": "^4.0.0"
```
so it shouldn't be a problem 🙂